In [ ]:
import pandas as pd

In [ ]:
train = pd.read_csv("/content/drive/MyDrive/LG_Aimers4_본선/train.csv")
test = pd.read_csv("/content/drive/MyDrive/LG_Aimers4_본선/submission.csv")

In [ ]:
# 각 피처별 고유값의 수를 계산합니다.
nunique_values = train.nunique()

# 35개 이하의 고유값을 가진 피처들을 필터링합니다.
features_with_few_unique_values = nunique_values[nunique_values <= 35].index.tolist()

# 타겟 변수 'is_converted'를 제외한 피처 리스트를 생성합니다.
features_to_encode = [feature for feature in features_with_few_unique_values if feature != 'is_converted']

# 각 피처에 대해 타겟 인코딩을 적용하고 원본 데이터프레임에 추가합니다.
for feature in features_to_encode:
    # 각 카테고리별로 'is_converted'의 평균값을 계산합니다.
    encoded_feature_name = f"{feature}_encoded"
    train[encoded_feature_name] = train.groupby(feature)['is_converted'].transform('mean')
# For each feature with 35 or fewer unique values, calculate value_counts()
value_counts_per_feature = {feature: train[feature].value_counts() for feature in features_to_encode}
value_counts_per_feature


{'bant_submit': bant_submit
 0.50    17192
 1.00    17061
 0.25    12852
 0.75    11681
 0.00      513
 Name: count, dtype: int64,
 'business_unit': business_unit
 ID          25563
 AS          24774
 IT           8664
 Solution      296
 CM              2
 Name: count, dtype: int64,
 'customer_type': customer_type
 End-Customer                    6647
 End Customer                    3996
 Specifier/ Influencer           2525
 Channel Partner                 1368
 Service Partner                  349
 Solution Eco-Partner             146
 Installer/Contractor              52
 Specifier / Influencer            43
 Corporate                         31
 HVAC Engineer                     23
 Engineer                          20
 Developer                         17
 Technician                        16
 Consultant                        15
 Other                             10
 Home Owner                        10
 Manager / Director                 8
 End-user                           

In [ ]:
# 중복 행을 제거하고 각 중복 그룹에서 첫 번째 행만 남김
train = train.drop_duplicates()
train = train.reset_index(drop = True)
train

bant_submit      expected_budget  \
0             0.75                  NaN   
1             0.25                  NaN   
2             0.25                  NaN   
3             1.00  $100,000 ~ $500,000   
4             0.50                  NaN   
...            ...                  ...   
57060         0.75                  NaN   
57061         0.75                  NaN   
57062         0.50                  NaN   
57063         0.50                  NaN   
57064         0.50     less_than_£7,000   

                                        customer_country business_unit  \
0                                 /DKI Jakarta/Indonesia            AS   
1                                /Minneola/United States            ID   
2      4880 Stevens Creek Blvd  Cloud IT/Hollywood/Un...            IT   
3                                     /Wiesbaden/Germany            ID   
4                                            //Singapore            ID   
...                                                  ...           ...   
57060                                       /Noida/India            ID   
57061                      /Sao Bernardo do Campo/Brazil            ID   
57062                                  /Alexandria/Egypt            AS   
57063                             /Mogi guaçu SP /Brazil            AS   
57064                                   //United Kingdom            IT   

       com_reg_ver_win_rate  customer_idx   lead_date          customer_type  \
0                       NaN         33931  2022-10-15           End-Customer   
1                  0.390244         42490  2023-05-31                    NaN   
2                       NaN         10350  2023-04-19   Solution Eco-Partner   
3                  0.017544         19750  2021-10-30                    NaN   
4                       NaN         45802  2023-04-11  Specifier/ Influencer   
...                     ...           ...         ...                    ...   
57060                   NaN         17289  2021-07-02                    NaN   
57061                   NaN         23686  2021-07-27                    NaN   
57062              0.049689         29370  2021-08-26                    NaN   
57063                   NaN         22609  2022-04-25                    NaN   
57064                   NaN         41443  2022-09-12                    NaN   

      customer_history  enterprise  ...  transfer_agreement  ver_cus  ver_pro  \
0             Existing  Enterprise  ...                   Y        0        0   
1                  New         SMB  ...                   N        0        0   
2                  New         SMB  ...                   Y        0        0   
3                  New         SMB  ...                   Y        0        0   
4                  New         SMB  ...                   Y        0        0   
...                ...         ...  ...                 ...      ...      ...   
57060         Existing  Enterprise  ...                   Y        0        0   
57061              New         SMB  ...                   Y        0        0   
57062         Existing  Enterprise  ...                   Y        0        0   
57063              New  Enterprise  ...                   Y        0        0   
57064              New         SMB  ...                   Y        0        0   

       ver_win_rate_x ver_win_rate_mean_upper ver_win_ratio_per_bu  \
0            0.000060                     1.0             0.128571   
1            0.000572                     0.0             0.048630   
2                 NaN                     NaN                  NaN   
3            0.000572                     0.0             0.048630   
4                 NaN                     NaN                  NaN   
...               ...                     ...                  ...   
57060             NaN                     NaN                  NaN   
57061             NaN                     NaN                  NaN   
57062        0.000298                

In [ ]:
train[train['is_converted'] == True]

bant_submit    expected_budget  \
1             0.25                NaN   
2             0.25                NaN   
23            0.75                NaN   
60            0.75                NaN   
65            0.25                NaN   
...            ...                ...   
59201         1.00  $10,000 ~ $50,000   
59236         0.50                NaN   
59238         0.50                NaN   
59248         0.25                NaN   
59249         0.25                NaN   

                                        customer_country business_unit  \
1                                /Minneola/United States            ID   
2      4880 Stevens Creek Blvd  Cloud IT/Hollywood/Un...            IT   
23                                           //Hong Kong            ID   
60                                    /Navi mumbai/India            ID   
65                                     /Peddapuram/India            AS   
...                                                  ...           ...   
59201                                  /Bangalore /India            ID   
59236                                           //Brazil            AS   
59238                               /jeddah/Saudi Arabia            AS   
59248  753 Forest Street  Medical/Marlborough/United ...            IT   
59249                                   /Salvador/Brazil            ID   

       com_reg_ver_win_rate  customer_idx   lead_date         customer_type  \
1                  0.390244         42490  2023-05-31                   NaN   
2                       NaN         10350  2023-04-19  Solution Eco-Partner   
23                      NaN         25096  2022-06-30                   NaN   
60                      NaN         32912  2023-04-08                   NaN   
65                      NaN         41227  2023-01-07          End-Customer   
...                     ...           ...         ...                   ...   
59201              0.057534         22552  2023-04-28          End-Customer   
59236                   NaN         25096  2022-05-31                   NaN   
59238                   NaN         37714  2023-04-01          End-Customer   
59248              0.642857          7576  2023-04-28          End Customer   
59249                   NaN         25096  2021-08-10                   NaN   

      customer_history  enterprise  ...  ver_win_rate_mean_upper  \
1                  New         SMB  ...                      0.0   
2                  New         SMB  ...                      NaN   
23                 New  Enterprise  ...                      NaN   
60                 New  Enterprise  ...                      NaN   
65                 New  Enterprise  ...                      NaN   
...                ...         ...  ...                      ...   
59201              New  Enterprise  ...                      0.0   
59236              New  Enterprise  ...                      NaN   
59238              New         SMB  ...                      NaN   
59248              New         SMB  ...                      1.0   
59249              New  Enterprise  ...                      NaN   

       ver_win_ratio_per_bu           business_area  business_subarea  \
1                  0.048630               education               NaN   
2                       NaN                     NaN               NaN   
23                      NaN                     NaN               NaN   
60                      NaN                     NaN               NaN   
65                      NaN                     NaN               NaN   
...                     ...                     ...               ...   
59201              0.064566      corporate / office      Construction   
59236                   NaN                     NaN               NaN   
59238                   NaN                     NaN               NaN   
59248                   NaN  hospital & health care            Others   
59249                   NaN                     NaN     

In [ ]:
#transfer_date

In [ ]:
train[['expected_budget','lead_date','lead_from_channel','lead_description','event_name','prefer_ver_count','prefer_ver_mean','transfer_agreement','ver_win_rate_mean_upper']]

expected_budget   lead_date lead_from_channel  \
0                      NaN  2022-10-15     lg.com(local)   
1                      NaN  2023-05-31    promotion page   
2                      NaN  2023-04-19               NaN   
3      $100,000 ~ $500,000  2021-10-30            lg.com   
4                      NaN  2023-04-11     lg.com(local)   
...                    ...         ...               ...   
57060                  NaN  2021-07-02            lg.com   
57061                  NaN  2021-07-27            lg.com   
57062                  NaN  2021-08-26    lg.com(global)   
57063                  NaN  2022-04-25   webinar(online)   
57064     less_than_£7,000  2022-09-12          facebook   

                                        lead_description  \
0                                                    NaN   
1         Are there any grants for technology available?   
2                                                    NaN   
3      Dear Ladies and Gentlemen, \n\nfor three of ou...   
4                                                   test   
...                                                  ...   
57060  Please share quotation for 80" 4k display. Ple...   
57061                                                NaN   
57062      how much is 5 zones ac \n1.5 hp x 4\n5 hp x 1   
57063                                                NaN   
57064                                                NaN   

                                             event_name  prefer_ver_count  \
0                                 AS_IN_LG.com_I2B_2022          0.076923   
1                            ID_US_Createboard_Campaign          0.095436   
2        IT_US_Cloud_IT_VMUG_UserCon_Event_Leads_230323               NaN   
3                                      ID_HQ_LG.com_I2B          0.095436   
4                                  ID_SL_LG.com_I2B_Dev               NaN   
...                                                 ...               ...   
57060                                  ID_IL_LG.com_I2B               NaN   
57061                                  ID_SP_LG.com_I2B               NaN   
57062                          AS_Global_M&A_LG.com_I2B          0.269231   
57063  AS_SP_Webinar_Integrated_(220424)_Live Esquadrão               NaN   
57064              IT_EU_RHQ_IFA_DualUp MNT_Lead Ads_FB               NaN   

       prefer_ver_mean transfer_agreement  ver_win_rate_mean_upper  
0             0.019648                  Y                      1.0  
1             0.067718                  N                      0.0  
2                  NaN                  Y                      NaN  
3             0.067718                  Y                      0.0  
4                  NaN                  Y                      NaN  
...                ...                ...                      ...  
57060              NaN                  Y                      NaN  
57061              NaN                  Y                      NaN  
57062         0.003494                  Y                      0.0  
57063              NaN                  Y                      NaN  
57064              NaN                  Y                      NaN  

[57065 rows x 9 columns]

### expected_budget
고객이 희망하는 예산 범위 (텍스트)

In [ ]:
train_expected_budget = train[['expected_budget']].copy()
test_expected_budget = test[['expected_budget']].copy()
train_expected_budget['expected_budget'].unique(), test_expected_budget['expected_budget'].unique()#0 ~ 1까지 0.25씩 있음

(array([nan, '$100,000 ~ $500,000', '$50,000 ~ $100,000',
        'More than $1,000,000', '$200,000 ~ $400,000', 'Less than $10,000',
        'less_than_$100,000', '$10,000 ~ $50,000', 'OnGoing',
        'Less than $100,000', '$500,000 ~ $1,000,000',
        '$100,000_-_$500,000', 'Less than $50,000',
        '500.000€__~__1.000.000€', 'more_than_$1,000,000',
        'More than $50,000', '$500,000_-_$1,000,000', 'Ongoing',
        '$200,000_~_$400,000', 'menos_de_10.000€', '$400,000 ~ $600,000',
        '$500,000 to less than $1M', '$100,000 ~ $200,000',
        'less_than_£7,000', '$100,000_~_$200,000', '€10,000_~_€50,000',
        'weniger_als_€10,000', '€100,000_~_€500,000', '€50,000_~_€100,000',
        'Menos de 10.000€', '$600,000 ~ $800,000', '$800,000 ~ $1,000,000',
        '25000 USD', 'On-going', 'more_than_£700,000',
        '$10M to less than $25M', '$10,000_~_$50,000', '$50.337.000 COP',
        '€100,000 ~ €500,000', '$400,000_~_$600,000',
        'More than $1.000.000', 

# lead_date
Lead 정보 생성일


In [ ]:
train_lead_date = train[['lead_date']].copy()
test_lead_date = test[['lead_date']].copy()
train_lead_date['lead_date'].unique(), test_lead_date['lead_date'].unique()#0 ~ 1까지 0.25씩 있음

(array(['2022-10-15', '2023-05-31', '2023-04-19', '2021-10-30',
        '2023-04-11', '2022-11-08', '2022-05-12', '2022-04-17',
        '2023-03-01', '2023-01-10', '2023-05-24', '2023-03-31',
        '2021-05-13', '2022-06-08', '2022-04-18', '2021-05-27',
        '2023-01-24', '2022-02-18', '2022-08-18', '2022-03-16',
        '2023-01-21', '2023-02-10', '2022-06-28', '2022-06-30',
        '2022-05-03', '2022-05-21', '2023-04-27', '2023-05-29',
        '2023-02-11', '2022-02-25', '2023-05-18', '2022-04-09',
        '2022-05-11', '2022-02-13', '2022-05-10', '2022-06-13',
        '2021-11-26', '2023-05-21', '2022-05-06', '2022-11-29',
        '2023-04-28', '2023-03-17', '2023-02-28', '2021-12-27',
        '2023-01-26', '2022-01-14', '2023-02-15', '2022-12-16',
        '2022-06-11', '2022-10-30', '2021-08-05', '2022-03-29',
        '2022-07-05', '2023-03-04', '2022-07-19', '2022-09-20',
        '2023-01-18', '2023-04-08', '2023-04-20', '2022-11-26',
        '2022-04-26', '2023-01-07', '202

### lead_from_channel
Lead 정보가 수집된 채널 이름 (lg.com, 페이스북, 웨비나 등 다양한 채널을 통해서 수집)

In [ ]:
train_lead_from_channel = train[['lead_from_channel']].copy()
test_lead_from_channel = test[['lead_from_channel']].copy()
train_lead_from_channel['lead_from_channel'].unique(), test_lead_from_channel['lead_from_channel'].unique()#0 ~ 1까지 0.25씩 있음

(array(['lg.com(local)', 'promotion page', nan, 'lg.com', 'lg.com(global)',
        'facebook', 'c-display', 'brochure & leaflet', 'webinar(online)',
        'microsite', 'youtube', 'instagram', 'webinar', 'gdn', 'linkedin',
        'newsletter(+blogsubscription)', 'exhibition', 'seminar', 'etc.',
        'roadshow', 'digital showroom', 'sem', 'website', 'dealertrip',
        'eloqua', 'sub(webpage)', 'conference', 'casa mollde',
        '[mql] fs_id_standard signage_20211025',
        'eu_es_id_202012w4_sem hotels', 'db(imported)', 'micrositio',
        'casa cor exhibition', '[mql] fs_id_hotel tv_20211011',
        'newsletter', 'db (imported)', 'offlineevent',
        'eu_es_id_202012w4_led campaign', 'equipotel event',
        '[mql]vh_lgsafepass_20210915',
        'id_bn_controlroomwhitepaperform_210707', 'lsvp',
        '[mql]vh_lgsafepass_20210908', '[mql]vh_lgsafepass_20210909',
        '[mql] fs_id_led signage_20211026', 'information.display@lge.pl',
        '[mql] fs_id_email

### lead_description
객이 작성한 Lead Description 텍스트

In [ ]:
train_lead_description = train[['lead_description']].copy()
test_lead_description = test[['lead_description']].copy()
train_lead_description['lead_description'].unique(), test_lead_description['lead_description'].unique()#0 ~ 1까지 0.25씩 있음

(array([nan, 'Are there any grants for technology available?',
        'Dear Ladies and Gentlemen, \n\nfor three of our grading suites, we need a quotation for three XXXXX UltraFine OLED Pro 65EP5G Display.\n\nIt would be nice, if you can get in touch with me as soon as possible. \n\nKind regards\nThorsten Rosche',
        ..., 'for 12 th physics classroom',
        'Please share quotation for 80" 4k display. Please share email id fo sales person to enable to share detailed specification and qty.',
        'how much is 5 zones ac \n1.5 hp x 4\n5 hp x 1'], dtype=object),
 array([nan, 'i have an old system need to be replaced',
        'WE WANT TO PURCHASE , WHICH WILL BE COMPATIBLE OF 2-3 INDOOR AND ONE OUTDOOR',
        ...,
        'Interested to purchase Multo split for home, we have to plan install total 8 Nos X 1.5 ton and 2 nos X 2 ton and 2 nos X 1 ton air conditioner, Kindly give a quotation for the multi split system',
        "Voltage and Harmonic issues caused by VFD's in ARU

### event_name
영업 활동이 있었던 마케팅 이벤트

In [ ]:
train_event_name = train[['event_name']].copy()
test_event_name = test[['event_name']].copy()
train_event_name['event_name'].unique(), test_event_name['event_name'].unique()

(array(['AS_IN_LG.com_I2B_2022', 'ID_US_Createboard_Campaign',
        'IT_US_Cloud_IT_VMUG_UserCon_Event_Leads_230323', ...,
        'IT_MS_LG.com_I2B_Cloud',
        'AS_PH_Seminar_GL_210617_LGUniqueSpecifications_Attendee_form',
        'IT_Showroom_Medical(Demo Req)'], dtype=object),
 array(['ID_US_Event_COCC Partners Expo_20230614', 'AS_IL_LG.com_I2B_2022',
        'AS_EG_facebookleadscampaign_May2023', 'ID_IL_LG.com_I2B',
        'AS_SP_Webinar_Integrated_SeminarDate(220322)_CasaConteúdo',
        'calendario_completo_septiembre_2023', 'ID_US_MAGNIT-Site Survey',
        'ID_HQ_Education_Lead Campaign_Facebook',
        'ID_MEAHQ Sales Contact QR Code_Aug2023',
        'IT_US_Event_SHI EUC Summit_20230620',
        'AS_EG_Multi V Residential Campaign_Facebook_2021',
        'AS_TK_LG.com_I2B_2022', 'AS_EG_LG.com_I2B',
        'AS_PH_Seminar_REGISTRATION_Incl_MQLs_form_September 2023',
        'AS_EG_LG.com_I2B_2022_Ara', 'AS_SP_LG.com_I2B_2022',
        'ID_MS_Promos_AGO', 'AS_SP

### prefer_ver_count
특정 Business Unit 기준으로 영업 전환 성공 샘플에 대해서 Vertical Level 1 비율

In [ ]:
train_prefer_ver_count = train[['prefer_ver_count']].copy()
test_prefer_ver_count = test[['prefer_ver_count']].copy()
train_prefer_ver_count['prefer_ver_count'].unique(), test_prefer_ver_count['prefer_ver_count'].unique()#0 ~ 1까지 0.25씩 있음

(array([0.07692308, 0.09543568,        nan, 0.14522822, 0.21161826,
        0.13043478, 0.28846154, 0.05809129, 0.26923077, 0.03846154,
        0.04347826, 0.01659751, 0.43478261, 0.        , 0.05769231,
        0.11538462, 0.21576763, 0.11618257, 0.07053942, 0.17391304,
        0.02489627, 0.01244813, 1.        , 0.01923077]),
 array([0.21161826, 0.05769231, 0.07692308,        nan, 0.14522822,
        0.21576763, 0.02489627, 0.09543568, 0.28846154, 0.03846154,
        0.        , 0.01659751, 0.11618257, 0.26923077, 0.07053942,
        0.13043478, 0.11538462, 0.05809129, 0.17391304, 0.01244813,
        0.43478261, 0.01923077, 0.04347826]))

### 0.434783이랑 0.043478이랑 밸류 차이는 거의 없는데 결과 차이는 정반대, 얘는 차라리 카테고리로 쓰는게 나음(cat_feature)

In [ ]:
conversion_summary = train.groupby('prefer_ver_count')['is_converted'].value_counts(normalize=True).unstack(fill_value=0).join(
    train.groupby('prefer_ver_count')['is_converted'].value_counts().unstack(fill_value=0), rsuffix='_count')
conversion_summary_with_sum = conversion_summary.assign(
    sum=lambda x: x['True_count'] + x['False_count']
).sort_values(by='sum', ascending=False)
conversion_summary_rearranged = conversion_summary_with_sum[['sum', 'True', 'False', 'True_count', 'False_count']]
conversion_summary_rearranged

is_converted       sum      True     False  True_count  False_count
prefer_ver_count                                                   
0.211618          2410  0.087552  0.912448         211         2199
0.145228          1972  0.061866  0.938134         122         1850
0.038462          1374  0.037118  0.962882          51         1323
0.269231          1355  0.017712  0.982288          24         1331
0.095436          1334  0.065967  0.934033          88         1246
0.288462          1251  0.030376  0.969624          38         1213
0.116183          1043  0.069990  0.930010          73          970
0.215768           941  0.105207  0.894793          99          842
0.000000           845  0.053254  0.946746          45          800
0.043478           829  0.003619  0.996381           3          826
0.016598           775  0.073548  0.926452          57          718
0.434783           745  0.370470  0.629530         276          469
0.115385           704  0.028409  0.971591          20          684
0.070539           533  0.067542  0.932458          36          497
0.058091           411  0.077859  0.922141          32          379
0.057692           377  0.055703  0.944297          21          356
0.173913           365  0.046575  0.953425          17          348
0.024896           209  0.066986  0.933014          14          195
0.076923           152  0.098684  0.901316          15          137
0.130435           143  0.048951  0.951049           7          136
0.019231            59  0.135593  0.864407           8           51
0.012448            43  0.279070  0.720930          12           31
1.000000            29  0.034483  0.965517           1           28

In [ ]:
train[(train['prefer_ver_count'] > 0.043477) & (train['prefer_ver_count'] < 0.043479)]

bant_submit    expected_budget customer_country business_unit  \
44             0.5                NaN           Mexico            IT   
155            0.5                NaN           Mexico            IT   
190            0.5                NaN           Mexico            IT   
206            0.5                NaN           Mexico            IT   
323            0.5                NaN           Mexico            IT   
...            ...                ...              ...           ...   
56655          0.5                NaN           France            IT   
56735          0.5                NaN           Mexico            IT   
56798          0.5                NaN           Mexico            IT   
56919          0.5                NaN           Mexico            IT   
56964          1.0  Less than $50,000       /Lima/Peru            IT   

       com_reg_ver_win_rate  customer_idx   lead_date customer_type  \
44                      NaN         43811  2021-12-27           NaN   
155                     NaN          4385  2021-12-27           NaN   
190                     NaN         19959  2021-12-27           NaN   
206                     NaN         21321  2021-12-27           NaN   
323                     NaN         33331  2021-12-27           NaN   
...                     ...           ...         ...           ...   
56655                   NaN         40855  2021-12-16           NaN   
56735                   NaN          6267  2021-12-27           NaN   
56798                   NaN         39432  2021-12-27           NaN   
56919                   NaN          9282  2021-12-27           NaN   
56964                   NaN         38563  2021-12-25           NaN   

      customer_history  enterprise  ...  transfer_agreement  ver_cus  ver_pro  \
44                 New  Enterprise  ...                   Y        0        0   
155                New  Enterprise  ...                   Y        0        0   
190                New  Enterprise  ...                   Y        0        0   
206                New  Enterprise  ...                   Y        0        0   
323                New  Enterprise  ...                   Y        0        0   
...                ...         ...  ...                 ...      ...      ...   
56655              New  Enterprise  ...                   Y        0        0   
56735              New  Enterprise  ...                   Y        0        0   
56798              New  Enterprise  ...                   Y        0        0   
56919              New  Enterprise  ...                   Y        0        0   
56964              New         SMB  ...                   Y        0        0   

       ver_win_rate_x ver_win_rate_mean_upper ver_win_ratio_per_bu  \
44           0.000717                     0.0                  NaN   
155          0.000717                     0.0                  NaN   
190          0.000717                     0.0                  NaN   
206          0.000717                     0.0                  NaN   
323          0.000717                     0.0                  NaN   
...               ...                     ...                  ...   
56655        0.000717                     0.0                  NaN   
56735        0.000717                     0.0                  NaN   
56798        0.000717                     0.0                  NaN   
56919        0.000717                     0.0                  NaN   
56964        0.000717                     0.0                  NaN   

               business_area  business_subarea lead_owner is_converted  
44     hotel & accommodation               NaN        816        False  
155    hotel & accommodation               NaN        816        False  
190    hotel & accommodation               NaN        816        False  
206    hotel & accommodation               NaN        816        False  
323    hotel & accommodation               NaN        816        False  
...                      ...         

In [ ]:
test[(test['prefer_ver_count'] > 0.043477) & (test['prefer_ver_count'] < 0.043479)]

id  bant_submit       expected_budget  \
1990   78753         1.00     Less than $10,000   
2190   37415         1.00   $100,000 ~ $200,000   
2336   78754         1.00     Less than $10,000   
2666   39159         1.00   $400,000 ~ $600,000   
4139   78752         1.00    $50,000 ~ $100,000   
4814   80266         0.75     Less than $10,000   
4925   80268         1.00   $200,000 ~ $400,000   
5049   80264         1.00     Less than $10,000   
11304  39158         0.25                   NaN   
14477  39157         0.75  More than $1,000,000   
20766  37416         1.00   $100,000 ~ $200,000   
21146  39156         1.00     Less than $10,000   

                               customer_country business_unit  \
1990                         / Berlin / Germany            IT   
2190                      / Uk / United Kingdom            IT   
2336                                /  / France            IT   
2666                      / Wiesbaden / Germany            IT   
4139                   / Melbourne  / Australia            IT   
4814                     / Düsseldorf / Germany            IT   
4925                      / Budapest  / Hungary            IT   
5049                               /  / Germany            IT   
11304  233 South Beaudry Avenue  Los Angeles CA            IT   
14477                       /  Sharjah  / U.A.E            IT   
20766                            / Uk / Germany            IT   
21146               /Oxfordshire/United Kingdom            IT   

       com_reg_ver_win_rate  customer_idx   lead_date          customer_type  \
1990                    NaN         36050  2023-07-19           End Customer   
2190                    NaN         25833  2023-10-15        Channel Partner   
2336                    NaN         26606  2023-08-01           End Customer   
2666                    NaN         26886  2023-10-21  Specifier/ Influencer   
4139                    NaN          2368  2023-07-20        Channel Partner   
4814                    NaN         27280  2023-08-03           End Customer   
4925                    NaN         14764  2023-07-15           End Customer   
5049                    NaN          4877  2023-08-01           End Customer   
11304                   NaN         25855  2023-09-27           End Customer   
14477                   NaN          2150  2023-07-29        Service Partner   
20766                   NaN         23337  2023-10-15        Channel Partner   
21146                   NaN         28583  2023-06-21           End Customer   

      customer_history  ... transfer_agreement  ver_cus  ver_pro  \
1990               New  ...                NaN        1        0   
2190               New  ...                NaN        0        0   
2336               New  ...                NaN        1        0   
2666               New  ...                NaN        0        0   
4139               New  ...                NaN        0        0   
4814          Existing  ...                NaN        0        0   
4925               New  ...                NaN        0        0   
5049               New  ...                NaN        0        0   
11304              New  ...                NaN        0        0   
14477              New  ...                NaN        0        0   
20766              New  ...                NaN        0        0   
21146              New  ...                  Y        0        0   

       ver_win_rate_x  ver_win_rate_mean_upper ver_win_ratio_per_bu  \
1990         0.000717                      0.0                  NaN   
2190         0.000717                      0.0                  NaN   
2336         0.000717                      0.0                  NaN   
2666         0.000026                      1.0                  NaN   
4139         0.000717                      0.0                  NaN   
4814         0.000026                      1.0                  NaN   
4925         0.000026                      1.0                  NaN   
5049  

### prefer_ver_mean
‘prefer_ver_count’ 와 동일한 기준으로 계산되며 샘플 개수가 아닌 이익 값의 비율

In [ ]:
# "prefer_ver_mean" 피처의 train 데이터셋과 test 데이터셋(여기서는 submission 데이터셋으로 가정)의 고유값 비교
train_prefer_ver_mean_unique = set(train['prefer_ver_mean'].unique())
submission_prefer_ver_mean_unique = set(test['prefer_ver_mean'].unique())

# 두 데이터셋에서 "prefer_ver_mean" 피처의 고유값이 모두 같은지 확인
are_values_identical = train_prefer_ver_mean_unique == submission_prefer_ver_mean_unique

are_values_identical, len(train_prefer_ver_mean_unique), len(submission_prefer_ver_mean_unique)

(False, 30, 29)

In [ ]:
train_prefer_ver_mean = train[['prefer_ver_mean']].copy()
test_prefer_ver_mean = test[['prefer_ver_mean']].copy()
train_prefer_ver_mean['prefer_ver_mean'].unique(), test_prefer_ver_mean['prefer_ver_mean'].unique()#0 ~ 1까지 0.25씩 있음

(array([0.01964817, 0.06771804,        nan, 0.10351139, 0.02775566,
        0.11086892, 0.01662405, 0.06828766, 0.00349357, 0.06792683,
        0.20827492, 0.13097958, 0.4456556 , 0.        , 0.00567704,
        0.00894398, 0.39132051, 0.04622018, 0.03888582, 0.04416258,
        0.04993932, 0.03271702, 0.4498716 , 0.07219541, 0.02576436,
        0.05213623, 0.10611456, 0.29066086, 1.        , 0.00377723]),
 array([0.02775566, 0.00567704, 0.01964817,        nan, 0.10351139,
        0.04622018, 0.10611456, 0.06771804, 0.01662405, 0.00894398,
        0.        , 0.03271702, 0.02576436, 0.03888582, 0.00349357,
        0.04993932, 0.04416258, 0.13097958, 0.11086892, 0.39132051,
        0.4498716 , 0.06828766, 0.07219541, 0.29066086, 0.4456556 ,
        0.00377723, 0.06792683, 0.20827492, 0.05213623]))

In [ ]:
conversion_summary = train.groupby('prefer_ver_mean')['is_converted'].value_counts(normalize=True).unstack(fill_value=0).join(
    train.groupby('prefer_ver_mean')['is_converted'].value_counts().unstack(fill_value=0), rsuffix='_count')
conversion_summary_with_sum = conversion_summary.assign(
    sum=lambda x: x['True_count'] + x['False_count']
).sort_values(by='sum', ascending=False)
conversion_summary_rearranged = conversion_summary_with_sum[['sum', 'True', 'False', 'True_count', 'False_count']]
conversion_summary_rearranged

is_converted      sum      True     False  True_count  False_count
prefer_ver_mean                                                   
0.027756         2410  0.087552  0.912448         211         2199
0.103511         1972  0.061866  0.938134         122         1850
0.003494         1355  0.017712  0.982288          24         1331
0.067718         1334  0.065967  0.934033          88         1246
0.016624         1251  0.030376  0.969624          38         1213
0.038886         1043  0.069990  0.930010          73          970
0.046220          941  0.105207  0.894793          99          842
0.000000          845  0.053254  0.946746          45          800
0.208275          789  0.002535  0.997465           2          787
0.008944          781  0.026889  0.973111          21          760
0.445656          745  0.370470  0.629530         276          469
0.391321          704  0.028409  0.971591          20          684
0.049939          533  0.067542  0.932458          36          497
0.068288          411  0.077859  0.922141          32          379
0.449872          398  0.060302  0.939698          24          374
0.044163          379  0.039578  0.960422          15          364
0.072195          325  0.033846  0.966154          11          314
0.025764          265  0.094340  0.905660          25          240
0.106115          209  0.066986  0.933014          14          195
0.005677          207  0.067633  0.932367          14          193
0.067927          195  0.030769  0.969231           6          189
0.110869          183  0.043716  0.956284           8          175
0.032717          170  0.041176  0.958824           7          163
0.019648          152  0.098684  0.901316          15          137
0.130980          131  0.129771  0.870229          17          114
0.003777           59  0.135593  0.864407           8           51
0.290661           43  0.279070  0.720930          12           31
0.052136           40  0.150000  0.850000           6           34
1.000000           29  0.034483  0.965517           1           28

In [ ]:
train[(train['prefer_ver_mean'] > 0.208274) & (train['prefer_ver_mean'] < 0.208276)]

bant_submit    expected_budget customer_country business_unit  \
44             0.5                NaN           Mexico            IT   
155            0.5                NaN           Mexico            IT   
190            0.5                NaN           Mexico            IT   
206            0.5                NaN           Mexico            IT   
323            0.5                NaN           Mexico            IT   
...            ...                ...              ...           ...   
56655          0.5                NaN           France            IT   
56735          0.5                NaN           Mexico            IT   
56798          0.5                NaN           Mexico            IT   
56919          0.5                NaN           Mexico            IT   
56964          1.0  Less than $50,000       /Lima/Peru            IT   

       com_reg_ver_win_rate  customer_idx   lead_date customer_type  \
44                      NaN         43811  2021-12-27           NaN   
155                     NaN          4385  2021-12-27           NaN   
190                     NaN         19959  2021-12-27           NaN   
206                     NaN         21321  2021-12-27           NaN   
323                     NaN         33331  2021-12-27           NaN   
...                     ...           ...         ...           ...   
56655                   NaN         40855  2021-12-16           NaN   
56735                   NaN          6267  2021-12-27           NaN   
56798                   NaN         39432  2021-12-27           NaN   
56919                   NaN          9282  2021-12-27           NaN   
56964                   NaN         38563  2021-12-25           NaN   

      customer_history  enterprise  ...  transfer_agreement  ver_cus  ver_pro  \
44                 New  Enterprise  ...                   Y        0        0   
155                New  Enterprise  ...                   Y        0        0   
190                New  Enterprise  ...                   Y        0        0   
206                New  Enterprise  ...                   Y        0        0   
323                New  Enterprise  ...                   Y        0        0   
...                ...         ...  ...                 ...      ...      ...   
56655              New  Enterprise  ...                   Y        0        0   
56735              New  Enterprise  ...                   Y        0        0   
56798              New  Enterprise  ...                   Y        0        0   
56919              New  Enterprise  ...                   Y        0        0   
56964              New         SMB  ...                   Y        0        0   

       ver_win_rate_x ver_win_rate_mean_upper ver_win_ratio_per_bu  \
44           0.000717                     0.0                  NaN   
155          0.000717                     0.0                  NaN   
190          0.000717                     0.0                  NaN   
206          0.000717                     0.0                  NaN   
323          0.000717                     0.0                  NaN   
...               ...                     ...                  ...   
56655        0.000717                     0.0                  NaN   
56735        0.000717                     0.0                  NaN   
56798        0.000717                     0.0                  NaN   
56919        0.000717                     0.0                  NaN   
56964        0.000717                     0.0                  NaN   

               business_area  business_subarea lead_owner is_converted  
44     hotel & accommodation               NaN        816        False  
155    hotel & accommodation               NaN        816        False  
190    hotel & accommodation               NaN        816        False  
206    hotel & accommodation               NaN        816        False  
323    hotel & accommodation               NaN        816        False  
...                      ...         

In [ ]:
test[(test['prefer_ver_mean'] > 0.208274) & (test['prefer_ver_mean'] < 0.208276)]

id  bant_submit      expected_budget          customer_country  \
1990   78753          1.0    Less than $10,000        / Berlin / Germany   
2190   37415          1.0  $100,000 ~ $200,000     / Uk / United Kingdom   
2336   78754          1.0    Less than $10,000               /  / France   
4139   78752          1.0   $50,000 ~ $100,000  / Melbourne  / Australia   
20766  37416          1.0  $100,000 ~ $200,000            / Uk / Germany   

      business_unit  com_reg_ver_win_rate  customer_idx   lead_date  \
1990             IT                   NaN         36050  2023-07-19   
2190             IT                   NaN         25833  2023-10-15   
2336             IT                   NaN         26606  2023-08-01   
4139             IT                   NaN          2368  2023-07-20   
20766            IT                   NaN         23337  2023-10-15   

         customer_type customer_history  ... transfer_agreement  ver_cus  \
1990      End Customer              New  ...                NaN        1   
2190   Channel Partner              New  ...                NaN        0   
2336      End Customer              New  ...                NaN        1   
4139   Channel Partner              New  ...                NaN        0   
20766  Channel Partner              New  ...                NaN        0   

       ver_pro  ver_win_rate_x  ver_win_rate_mean_upper ver_win_ratio_per_bu  \
1990         0        0.000717                      0.0                  NaN   
2190         0        0.000717                      0.0                  NaN   
2336         0        0.000717                      0.0                  NaN   
4139         0        0.000717                      0.0                  NaN   
20766        0        0.000717                      0.0                  NaN   

               business_area business_subarea  lead_owner is_converted  
1990   hotel & accommodation              NaN         428          NaN  
2190   hotel & accommodation    Casino Resort         431          NaN  
2336   hotel & accommodation              NaN         474          NaN  
4139   hotel & accommodation            Hotel         390          NaN  
20766  hotel & accommodation           Cruise         863          NaN  

[5 rows x 40 columns]

### transfer_agreement
Lead 정보 국외 반출 동의 여부

In [ ]:
train_transfer_agreement = train[['transfer_agreement']].copy()
test_transfer_agreement = test[['transfer_agreement']].copy()
train_transfer_agreement['transfer_agreement'].unique(), test_transfer_agreement['transfer_agreement'].unique()#0 ~ 1까지 0.25씩 있음

(array(['Y', 'N', nan], dtype=object), array([nan, 'Y', 'N'], dtype=object))

In [ ]:
train_transfer_agreement['transfer_agreement'].value_counts(), test_transfer_agreement['transfer_agreement'].value_counts()

(transfer_agreement
 Y    47462
 N     9073
 Name: count, dtype: int64,
 transfer_agreement
 Y    2599
 N     225
 Name: count, dtype: int64)

In [ ]:
conversion_summary = train.groupby('transfer_agreement')['is_converted'].value_counts(normalize=True).unstack(fill_value=0).join(
    train.groupby('transfer_agreement')['is_converted'].value_counts().unstack(fill_value=0), rsuffix='_count')
conversion_summary_with_sum = conversion_summary.assign(
    sum=lambda x: x['True_count'] + x['False_count']
).sort_values(by='sum', ascending=False)
conversion_summary_rearranged = conversion_summary_with_sum[['sum', 'True', 'False', 'True_count', 'False_count']]
conversion_summary_rearranged

is_converted          sum      True     False  True_count  False_count
transfer_agreement                                                    
Y                   47462  0.091715  0.908285        4353        43109
N                    9073  0.031853  0.968147         289         8784

### ver_win_rate_mean_upper
Ratio_oppty – Ratio 값을 구해서 Vertical 별 평균 값보다 클 경우 1, 아니면 0 값으로 표현

In [ ]:
train_ver_win_rate_mean_upper = train[['ver_win_rate_mean_upper']].copy()
test_ver_win_rate_mean_upper = test[['ver_win_rate_mean_upper']].copy()
train_ver_win_rate_mean_upper['ver_win_rate_mean_upper'].unique(), test_ver_win_rate_mean_upper['ver_win_rate_mean_upper'].unique()#0 ~ 1까지 0.25씩 있음

(array([ 1.,  0., nan]), array([ 0.,  1., nan]))

In [ ]:
train_ver_win_rate_mean_upper['ver_win_rate_mean_upper'].value_counts(), test_ver_win_rate_mean_upper['ver_win_rate_mean_upper'].value_counts()

(ver_win_rate_mean_upper
 0.0    14307
 1.0     3592
 Name: count, dtype: int64,
 ver_win_rate_mean_upper
 0.0    6126
 1.0    2121
 Name: count, dtype: int64)

In [ ]:
conversion_summary = train.groupby('ver_win_rate_mean_upper')['is_converted'].value_counts(normalize=True).unstack(fill_value=0).join(
    train.groupby('ver_win_rate_mean_upper')['is_converted'].value_counts().unstack(fill_value=0), rsuffix='_count')
conversion_summary_with_sum = conversion_summary.assign(
    sum=lambda x: x['True_count'] + x['False_count']
).sort_values(by='sum', ascending=False)
conversion_summary_rearranged = conversion_summary_with_sum[['sum', 'True', 'False', 'True_count', 'False_count']]
conversion_summary_rearranged

is_converted               sum      True     False  True_count  False_count
ver_win_rate_mean_upper                                                    
0.0                      14307  0.053820  0.946180         770        13537
1.0                       3592  0.139198  0.860802         500         3092

### customer_history

In [ ]:
train_customer_history = train[['customer_history']].copy()
test_customer_history = test[['customer_history']].copy()
train_customer_history['customer_history'].unique(), test_customer_history['customer_history'].unique()#0 ~ 1까지 0.25씩 있음

(array(['Existing', 'New'], dtype=object),
 array(['New', 'Existing'], dtype=object))

In [ ]:
train.drop(['expected_budget','lead_date','lead_from_channel','lead_description','event_name','prefer_ver_count','prefer_ver_mean','transfer_agreement','ver_win_rate_mean_upper'],axis=1,inplace=True)

In [ ]:
train.to_csv('train_aim_check.csv',index=False)

In [ ]:
train

bant_submit                                   customer_country  \
0             0.75                             /DKI Jakarta/Indonesia   
1             0.25                            /Minneola/United States   
2             0.25  4880 Stevens Creek Blvd  Cloud IT/Hollywood/Un...   
3             1.00                                 /Wiesbaden/Germany   
4             0.50                                        //Singapore   
...            ...                                                ...   
59294         0.75                                       /Noida/India   
59295         0.75                      /Sao Bernardo do Campo/Brazil   
59296         0.50                                  /Alexandria/Egypt   
59297         0.50                             /Mogi guaçu SP /Brazil   
59298         0.50                                   //United Kingdom   

      business_unit  com_reg_ver_win_rate  customer_idx  \
0                AS                   NaN         33931   
1                ID              0.390244         42490   
2                IT                   NaN         10350   
3                ID              0.017544         19750   
4                ID                   NaN         45802   
...             ...                   ...           ...   
59294            ID                   NaN         17289   
59295            ID                   NaN         23686   
59296            AS              0.049689         29370   
59297            AS                   NaN         22609   
59298            IT                   NaN         41443   

               customer_type customer_history  enterprise  \
0               End-Customer         Existing  Enterprise   
1                        NaN              New         SMB   
2       Solution Eco-Partner              New         SMB   
3                        NaN              New         SMB   
4      Specifier/ Influencer              New         SMB   
...                      ...              ...         ...   
59294                    NaN         Existing  Enterprise   
59295                    NaN              New         SMB   
59296                    NaN         Existing  Enterprise   
59297                    NaN              New  Enterprise   
59298                    NaN              New         SMB   

       historical_existing_cnt  id_strategic_ver  ...  response_corporate  \
0                          4.0               NaN  ...               LGEIN   
1                          NaN               NaN  ...               LGEUS   
2                          NaN               NaN  ...               LGEUS   
3                          NaN               NaN  ...               LGEDG   
4                          NaN               NaN  ...               LGESL   
...                        ...               ...  ...                 ...   
59294                      3.0               NaN  ...               LGEIL   
59295                      NaN               NaN  ...               LGESP   
59296                      1.0               NaN  ...               LGEEG   
59297                      NaN               NaN  ...               LGESP   
59298                      NaN               NaN  ...               LGEUK   

         expected_timeline ver_cus  ver_pro ver_win_rate_x  \
0       less than 3 months       0        0       0.000060   
1                      NaN       0        0       0.000572   
2                      NaN       0        0            NaN   
3       less than 3 months       0        0       0.000572   
4       less than 3 months       0        0            NaN   
...                    ...     ...      ...            ...   
59294  3 months ~ 6 months       0        0            NaN   
59295  3 months ~ 6 months       0        0            NaN   
59296                  NaN       0        0       0.000298   
59297                  NaN       0        0            NaN   
59298                  NaN       0        0            NaN   

      ver_win_rati

## old prepro

In [ ]:
train = pd.read_csv("/content/drive/MyDrive/LG_Aimers4_본선/train_old_prepro.csv")
test = pd.read_csv("/content/drive/MyDrive/LG_Aimers4_본선/test_old_prepro.csv")
train

bant_submit  expected_budget  customer_country  business_unit  \
0             0.75               67                74              0   
1             0.25               67               166              2   
2             0.25               67               166              3   
3             1.00                4                60              2   
4             0.50               67               141              2   
...            ...              ...               ...            ...   
55284         0.75               67                73              2   
55285         0.75               67                26              2   
55286         0.50               67                50              0   
55287         0.50               67                26              0   
55288         0.50               79               165              3   

       com_reg_ver_win_rate  customer_idx  lead_date  customer_type  \
0                       NaN         26590        566              8   
1                  0.390244         36098        793             25   
2                       NaN           393        751             20   
3                  0.017544         10833        223             25   
4                       NaN         39777        743             21   
...                     ...           ...        ...            ...   
55284                   NaN          8098        105             25   
55285                   NaN         15206        129             25   
55286              0.049689         21522        158             25   
55287                   NaN         14010        398             25   
55288                   NaN         34935        534             25   

       customer_history  enterprise  ...   ratio_rate  log_ratio_rate  \
0                     0           0  ...  2127.245407        7.663053   
1                     1           1  ...    85.023333        4.454619   
2                     1           1  ...     0.000000        0.000000   
3                     1           1  ...    85.023333        4.454619   
4                     1           1  ...     0.000000        0.000000   
...                 ...         ...  ...          ...             ...   
55284                 0           0  ...     0.000000        0.000000   
55285                 1           1  ...     0.000000        0.000000   
55286                 0           0  ...    67.448952        4.226088   
55287                 1           0  ...     0.000000        0.000000   
55288                 1           1  ...     0.000000        0.000000   

            gob1          gob2         gob3   gob4       gob5      gob6  \
0            NaN  63817.362216  8508.981629  120.0        NaN       NaN   
1      33.179837   3911.073326          NaN    NaN  17.951220       NaN   
2            NaN      0.000000          NaN    NaN        NaN       NaN   
3       1.491637  19725.413298          NaN    NaN   4.070175       NaN   
4            NaN      0.000000          NaN    NaN        NaN       NaN   
...          ...           ...          ...    ...        ...       ...   
55284        NaN      0.000000     0.000000  393.0        NaN       NaN   
55285        NaN      0.000000          NaN    NaN        NaN       NaN   
55286   3.351501   2900.304957    67.448952   43.0   2.136646  0.049689   
55287        NaN      0.000000          NaN    NaN        NaN       NaN   
55288        NaN      0.000000          NaN    NaN        NaN       NaN   

       cluster_1  cluster_2  
0              2          2  
1              2          2  
2              2          0  
3              1          1  
4              2          0  
...          ...        ...  
55284          1          1  
55285          2          2  
55286          2          2  
55287          2          2  
55288          2          0  

[55289 rows x 104 columns]

In [ ]:
# Simplified text preprocessing function
def simplify_preprocess_text(text):
    # Convert text to lowercase and remove non-letter characters
    text = re.sub(r'[^a-zA-Z]', ' ', text.lower())
    # Remove extra spaces
    text = re.sub(r'\s+', ' ', text).strip()
    return text

# Apply the simplified preprocessing to the 'lead_description'
train['lead_description_clean_simple'] = train['lead_description'].astype(str).apply(simplify_preprocess_text)

# Vectorizing the preprocessed text with TF-IDF, simplified version
tfidf_vectorizer_simple = TfidfVectorizer(max_features=1000)  # Keeping the features to 1000 for demonstration
tfidf_matrix_simple = tfidf_vectorizer_simple.fit_transform(train['lead_description_clean_simple'])

tfidf_matrix_simple.shape


(59299, 1000)

In [ ]:
tfidf_matrix_simple

<59299x1000 sparse matrix of type '<class 'numpy.float64'>'
	with 311586 stored elements in Compressed Sparse Row format>

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
import re
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
import nltk

# Downloading NLTK resources (stopwords, punkt for tokenization, wordnet for lemmatization)
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

# Define a function for text preprocessing
def preprocess_text(text):
    # Remove HTML tags
    text = re.sub(r'<.*?>', '', text)
    # Remove non-letters and lower case all words
    text = re.sub(r'[^a-zA-Z]', ' ', text).lower()
    # Tokenize
    words = word_tokenize(text)
    # Remove stopwords
    words = [w for w in words if w not in stopwords.words('english')]
    # Lemmatize
    lemmatizer = WordNetLemmatizer()
    words = [lemmatizer.lemmatize(w) for w in words]
    # Join the words back into one string
    return ' '.join(words)

# Applying the preprocessing function to the 'lead_description' column
train['lead_description_clean'] = train['lead_description'].astype(str).apply(preprocess_text)

# Vectorizing the preprocessed text
tfidf_vectorizer = TfidfVectorizer(max_features=1000)  # Limiting to 1000 features for demonstration
tfidf_matrix = tfidf_vectorizer.fit_transform(train['lead_description_clean'])

tfidf_matrix.shape


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


(59299, 1000)

In [ ]:
train['lead_description_clean'].to_csv("aaaa.csv",index=False)

In [ ]:
# TF-IDF 벡터를 데이터 프레임으로 변환
tfidf_df = pd.DataFrame(tfidf_matrix_simple.toarray(), columns=tfidf_vectorizer_simple.get_feature_names_out())

# 기존 데이터 프레임과 결합하기 전에, 인덱스를 맞춰주어야 합니다. 이 경우 train의 인덱스를 사용합니다.
tfidf_df.index = train.index

# 기존의 데이터 프레임과 TF-IDF 데이터 프레임을 결합합니다. 여기서는 단순화를 위해 몇 개의 수치형 피처만 선택하여 결합해 보겠습니다.
# 실제로는 모델링에 필요한 모든 수치형 피처를 포함시켜야 합니다.
selected_numerical_features = train.select_dtypes(include=['int64', 'float64']).drop(columns=['customer_idx'])
combined_df = pd.concat([selected_numerical_features, tfidf_df], axis=1)

combined_df

bant_submit  com_reg_ver_win_rate  historical_existing_cnt  \
0             0.75                   NaN                      4.0   
1             0.25              0.390244                      NaN   
2             0.25                   NaN                      NaN   
3             1.00              0.017544                      NaN   
4             0.50                   NaN                      NaN   
...            ...                   ...                      ...   
59294         0.75                   NaN                      3.0   
59295         0.75                   NaN                      NaN   
59296         0.50              0.049689                      1.0   
59297         0.50                   NaN                      NaN   
59298         0.50                   NaN                      NaN   

       id_strategic_ver  it_strategic_ver  idit_strategic_ver  \
0                   NaN               NaN                 NaN   
1                   NaN               NaN                 NaN   
2                   NaN               NaN                 NaN   
3                   NaN               NaN                 NaN   
4                   NaN               NaN                 NaN   
...                 ...               ...                 ...   
59294               NaN               NaN                 NaN   
59295               NaN               NaN                 NaN   
59296               NaN               NaN                 NaN   
59297               NaN               NaN                 NaN   
59298               NaN               NaN                 NaN   

       lead_desc_length  prefer_ver_count  prefer_ver_mean  ver_cus  ...   xe  \
0                    30          0.076923         0.019648        0  ...  0.0   
1                    46          0.095436         0.067718        0  ...  0.0   
2                     3               NaN              NaN        0  ...  0.0   
3                   232          0.095436         0.067718        0  ...  0.0   
4                     4               NaN              NaN        0  ...  0.0   
...                 ...               ...              ...      ...  ...  ...   
59294               131               NaN              NaN        0  ...  0.0   
59295                31               NaN              NaN        0  ...  0.0   
59296                43          0.269231         0.003494        0  ...  0.0   
59297                22               NaN              NaN        0  ...  0.0   
59298                 3               NaN              NaN        0  ...  0.0   

       xxx     xxxxx  xxxxxot  xxxxxpping  year  years  yes       you  your  
0      0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
1      0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
2      0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
3      0.0  0.118307      0.0         0.0   0.0    0.0  0.0  0.125126   0.0  
4      0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
...    ...       ...      ...         ...   ...    ...  ...       ...   ...  
59294  0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
59295  0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
59296  0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
59297  0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  
59298  0.0  0.000000      0.0         0.0   0.0    0.0  0.0  0.000000   0.0  

[59299 rows x 1015 columns]

In [ ]:

train['event_name_first_part'] = train['event_name'].apply(lambda x: x.split('_', 1)[0] if pd.notnull(x) else x)
def extract_second_part(event_name):
    if pd.isnull(event_name):
        return event_name
    parts = event_name.split('_', 1)
    if len(parts) > 1:
        second_part = re.split('_| ', parts[1], 1)[0]
        return second_part
    return None

train['event_name_second_part'] = train['event_name'].apply(extract_second_part)


event_name event_name_first_part  \
0                           AS_IN_LG.com_I2B_2022                    AS   
1                      ID_US_Createboard_Campaign                    ID   
2  IT_US_Cloud_IT_VMUG_UserCon_Event_Leads_230323                    IT   
3                                ID_HQ_LG.com_I2B                    ID   
4                            ID_SL_LG.com_I2B_Dev                    ID   

                        event_name_second_part  
0                           IN_LG.com_I2B_2022  
1                      US_Createboard_Campaign  
2  US_Cloud_IT_VMUG_UserCon_Event_Leads_230323  
3                                HQ_LG.com_I2B  
4                            SL_LG.com_I2B_Dev

In [ ]:
def extract_second_part(event_name):
    if pd.isnull(event_name):
        return event_name
    parts = event_name.split('_', 1)
    if len(parts) > 1:
        second_part = re.split('_| ', parts[1], 1)[0]
        return second_part
    return None

train['event_name_second_part'] = train['event_name'].apply(extract_second_part)

train[['event_name', 'event_name_first_part', 'event_name_second_part']].to_csv("Checking_event.csv",index=False)